### PR TITLE
Fix double-width characters disappearing when wrapping

### LINKED ISSUE #3176: [BUG] Chunks of text go missing when writing Asian text (wrapping issue)
### Description

(Originally reported in Textual: https://github.com/Textualize/textual/issues/3567)

When you print Asian text (specifically Chinese and Japanese, which do not use spaces), portions of the text go missing, making it unreadable.

This seems to be related to wrapping, as the characters which go missing are at the end of a line.
Instead of being wrapped on to a new line, they disappear.

There are [rules for wrapping](https://en.m.wikipedia.org/wiki/Line_breaking_rules_in_East_Asian_languages) in these languages which would take more effort to adhere to, but at the very least, text should not go missing when printed.

### Examples

For example, running the snippet below, the `7` in `1670` disappears:

```python
from rich.console import Console
console = Console(width=20)
console.print("アプリケーションは1670万色を使用でき")
```

Output:

```
アプリケーションは16
0万色を使用でき
```

And in many cases, like those reported in this issue, multiple characters disappear:

```python
from rich.console import Console
console = Console(width=20)
console.print("TextualはPythonの高速アプリケーション開発フレームワークです")
```

Output:

```
TextualはPy
thonの高速アプリ
ケーション開発フレー
```

Notice that many of the characters at the end of the text are completely missing (`ムワークです`).

### LINKED ISSUE #3567: Incorrect word wrapping for Static widget with Chinese / Japanese characters
## Description
Word wrapping for Static widget seems not working properly with Chinese and Japanese characters.

Please continue reading for more details.

## Reproducible Example
```Python
from textual.app import App, ComposeResult
from textual.containers import Vertical
from textual.widgets import Static


class ReproApp(App):
    CSS = """
    .view {
        width: 100%;
        height: auto;
        margin: 1;
        border: heavy green;
    }
    """

    ENG_TEXT = "Textual is a Rapid Application Development framework for Python. Build sophisticated user interfaces with a simple Python API. Run your apps in the terminal and (coming soon) a web browser! Textual adds interactivity to Rich with an API inspired by modern web development. On modern terminal software (installed by default on most systems), Textual apps can use 16.7 million colors with mouse support and smooth flicker-free animation. A powerful layout engine and re-usable components makes it possible to build apps that rival the desktop and web experience. Textual runs on Linux, macOS, and Windows. Textual requires Python 3.7 or above. The Textual repository comes with a number of examples you can experiment with or use as a template for your own projects. Join the Textual developers and community on our Discord Server. Head over to the Textual documentation to start building!"
    CHS_TEXT = "Textual是Python的一个快速应用程序开发框架。使用简单的Python API构建复杂的用户界面。在终端和（即将推出的）网络浏览器中运行您的应用程序！Textual通过受现代web开发启发的API为Rich添加了交互性。在现代终端软件（大多数系统默认安装）上，纹理应用程序可以使用1670万种颜色，并支持鼠标和平滑无闪烁的动画。强大的布局引擎和可重复使用的组件使构建与桌面和网络体验相媲美的应用程序成为可能。Textual在Linux、macOS和Windows上运行。Textual需要Python 3.7或更高版本。Textural存储库提供了许多示例，您可以对其进行实验，也可以将其用作自己项目的模板。加入我们Discord服务器上的Textual开发人员和社区。前往文本文档开始构建！"
    JPN_TEXT = "TextualはPythonの高速アプリケーション開発フレームワークです。単純なPython APIを使用して複雑なユーザーインタフェースを構築します。端末および（近日公開予定の）Webブラウザでアプリケーションを実行します！Textualは、現代のweb開発にヒントを得たAPIを通じてRichにインタラクティブ性を追加した。現代のターミナルソフトウェア（ほとんどのシステムのデフォルトインストール）では、テクスチャアプリケーションは1670万色を使用でき、マウスと滑らかでフリッカーのないアニメーションをサポートします。強力なレイアウトエンジンと再利用可能なコンポーネントにより、デスクトップやネットワーク体験に匹敵するアプリケーションの構築が可能になります。TextualはLinux、macOS、Windowsで動作します。TextualにはPython 3.7以上のバージョンが必要です。Texturalリポジトリには多くの例があり、実験を行うことも、独自のプロジェクトのテンプレートとして使用することもできます。Dell Discordサーバ上のTextual開発者とコミュニティに参加します。文書ドキュメントの構築を開始します！"
    KOR_TEXT = "Textual은 Python의 빠른 응용 프로그램 개발 프레임워크입니다.간단한 Python API를 사용하여 복잡한 사용자 인터페이스를 구축합니다.터미널 및 (출시 예정) 웹 브라우저에서 응용 프로그램을 실행하십시오!Textual은 현대 웹 개발에서 영감을 받은 API를 통해 Rich에 상호 작용을 추가했습니다.현대 터미널 소프트웨어 (대부분의 시스템 기본 설치) 에서 텍스쳐 응용 프로그램은 1670 만 가지 색상을 사용할 수 있으며 마우스와 깜박이지 않는 매끄러운 애니메이션을 지원합니다.강력한 레이아웃 엔진과 재사용 가능한 구성 요소로 데스크탑 및 네트워크 환경에 필적하는 어플리케이션을 구축할 수 있습니다.Textual은 Linux, macOS 및 Windows에서 실행됩니다.Textual에는 Python 3.7 이상이 필요합니다.Textural 저장소는 실험하거나 프로젝트의 템플릿으로 사용할 수 있는 많은 예를 제공합니다.Discord 서버의 Textual 개발자 및 커뮤니티에 참여하십시오.텍스트 문서로 이동하여 구문 시작!"

    def compose(self) -> ComposeResult:
        yield Vertical(
            Static(self.ENG_TEXT, classes="view"),
            Static(self.CHS_TEXT, classes="view"),
            Static(self.JPN_TEXT, classes="view"),
            Static(self.KOR_TEXT, classes="view"),
        )


def main():
    app = ReproApp()
    app.run()
```

Preview

Please pay attention to the end of the second line and the beginning of the third line in the Chinese paragraph. The sentence should be `Textual通过受现代web开发启发的API为Rich添加了交互性。在现代终端软件（大多数系统默认安装）上，纹理应用程序可以使用1670万种颜色，并支持鼠标和平滑无闪烁的动画。`, however the part `加了交互性。在现代` is missing in the actual presentation.

Multiple lines in the same paragraph are with the similar issue. So for the Japanese paragraph below. Given that this feature works for both Korean and English, I guess that this may have something to do with the fact that there are hardly any spaces in a common Chinese or Japanese sentence.

#### Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.40.0 |
| Rich    | 13.6.0 |

#### Python

| Name           | Value                                                                                           |
|----------------|-------------------------------------------------------------------------------------------------|
| Version        | 3.8.0                                                                                           |
| Implementation | CPython                                                                                         |
| Compiler       | MSC v.1916 64 bit (AMD64)                                                                       |
| Executable     | C:\Users\Max\AppData\Local\pypoetry\Cache\virtualenvs\clipper-VCLAtNFH-py3.8\Scripts\python.exe |

#### Operating System

| Name    | Value      |
|---------|------------|
| System  | Windows    |
| Release | 10         |
| Version | 10.0.19041 |

#### Terminal

| Name                 | Value            |
|----------------------|------------------|
| Terminal Application | Windows Terminal |
| TERM                 | *Not set*        |
| COLORTERM            | *Not set*        |
| FORCE_COLOR          | *Not set*        |
| NO_COLOR             | *Not set*        |

#### Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=121, height=30 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 121                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 30                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |

### PR DESCRIPTION
**Pull request by darrenburns**
## Type of changes

- [x] Bug fix
- [ ] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [x] I've run the latest [black](https://github.com/psf/black) with default args on new code.
- [x] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [x] I've added tests for new code.
- [x] I accept that @willmcgugan may be pedantic in the code review.

## Description

Update wrapping logic to fix issues with CJK charcters disappearing when the "fold" location sat *within* a double-width character. Ensure we retain browser logic of: 

> if there is no space on the current line, move to a new line, and if theres not enough space on the entire new line, fold the text over multiple lines at appropriate locations.

Adds some additional tests and docstrings, documentation etc.

Fixes #3176 

The wrapping process is overall still quite simple and doesn't match the browser in many cases. For example, wrapping does not consider punctuation (lines can begin with punctuation), and whitespace is handled differently (but practically speaking it seems sensible).

### CODE REVIEW

File: rich/_wrap.py [hunk 1]
```diff
@@ -17,40 +20,71 @@ def words(text: str) -> Iterable[Tuple[int, int, str]]:
...
+        width: The available cell width.
+        fold: If True, words longer than `width` will be folded onto a new line.
+
+    Returns:
+        A list of indices to break the line at.
+    """
+    break_positions: list[int] = []  # offsets to insert the breaks at
+    append = break_positions.append
+    cell_offset = 0
     _cell_len = cell_len
```
Comment (rodrigogiraoserrao) [inline rich/_wrap.py:39]: Is this a micro optimisation to bring `cell_len` into the current scope?

Comment (darrenburns) [reply rich/_wrap.py:39]: Yes, it means that when we call it it is retrieved from the local scope, which is faster.

File: rich/_wrap.py [hunk 2]
```diff
@@ -17,40 +20,71 @@ def words(text: str) -> Iterable[Tuple[int, int, str]]:
...
+
     for start, _end, word in words(text):
         word_length = _cell_len(word.rstrip())
-        if line_position + word_length > width:
+        remaining_space = width - cell_offset
+        word_fits_remaining_space = remaining_space - word_length >= 0
```
Comment (rodrigogiraoserrao) [inline rich/_wrap.py:44]: Why didn't you make the comparison directly?

```suggestion
        word_fits_remaining_space = remaining_space >= word_length
```

(Which could also go directly in the `if` statement, as `if remaining_space >= word_length`...)

File: rich/_wrap.py [hunk 3]
```diff
@@ -17,40 +20,71 @@ def words(text: str) -> Iterable[Tuple[int, int, str]]:
         word_match = re_word.match(text, end)
 
 
-def divide_line(text: str, width: int, fold: bool = True) -> List[int]:
-    divides: List[int] = []
-    append = divides.append
-    line_position = 0
+def divide_line(text: str, width: int, fold: bool = True) -> list[int]:
```
Comment (rodrigogiraoserrao) [inline rich/_wrap.py:23]: I'm assuming `cell_offset` is keeping track of how much space is already occupied in the line we're currently filling, right?
I _personally_ don't love the name but maybe that could be mitigated with a short comment next to it?

File: rich/_wrap.py [hunk 4]
```diff
@@ -1,13 +1,16 @@
...
+from typing import Iterable
 
 from ._loop import loop_last
 from .cells import cell_len, chop_cells
 
 re_word = re.compile(r"\s*\S+\s*")
 
 
-def words(text: str) -> Iterable[Tuple[int, int, str]]:
+def words(text: str) -> Iterable[tuple[int, int, str]]:
```
Comment (rodrigogiraoserrao) [inline rich/_wrap.py:12]: The first time I read the function name & docstring, I immediately expected the function to take a string like

```py
"Hey    how     are   you"
#0123456789012345678901234
```

And produce something like `(0, 2, "Hey"), (7, 9, "how"), ...` so I didn't even bother checking the regex pattern.
This, in turn, made it confusing to read `divide_line`.
So, all I'm saying is that maybe “Yields each word” is a bad start of the docstring because I don't know many people that would call the string `"Hey    "` a word.

Maybe mention that surrounding whitespace also gets consumed?
Or reword it to say that `words` breaks a string on non-whitespace characters or something like that?

Not sure if you get what I'm trying to say.

File: rich/text.py [hunk 5]
```diff
@@ -1204,7 +1204,7 @@ def wrap(
 
         Args:
             console (Console): Console instance.
-            width (int): Number of characters per line.
+            width (int): Number of single-width characters per line.
```
Comment (rodrigogiraoserrao) [inline rich/text.py:1207]: Technically correct but a bit awkward, no?
How about the suggestion below or something of the sorts?

```suggestion
            width (int): Number of cells available per line.
```

File: tests/test_cells.py [hunk 6]
```diff
@@ -40,3 +41,21 @@ def test_set_cell_size_infinite():
...
+def test_chop_cells_double_width_boundary():
+    """The available width lies within a double-width character."""
+    text = "ありがとう"
+    assert chop_cells(text, 3) == ["あ", "り", "が", "と", "う"]
+
+
+def test_chop_cells_mixed_width():
+    """Mixed single and double-width characters."""
+    text = "あ1り2が3と4う56"
+    assert chop_cells(text, 3) == ["あ1", "り2", "が3", "と4", "う5", "6"]
```
Comment (rodrigogiraoserrao) [inline tests/test_cells.py:61]: I mixed it up a bit more to be less regular.
(I did this by hand and didn't run the test after! I also just copied and pasted some characters, maybe there was some meaning in here that I butchered... 😬)

```suggestion
    text = "あ1り2がり34とりう56"
    assert chop_cells(text, 3) == ["あ1", "り2", "が", "り3", "4と", "り", "う5", "6"]
```

**Review by rodrigogiraoserrao — APPROVED** [review]
I suggested a couple of tweaks but they are fairly subjective, so I won't block the merge because of them.

File: tests/test_cells.py (hunk 6 above)
Comment (darrenburns) [reply tests/test_cells.py]: You turned "arigatou" into "arigaritoriu" 😄

**Review by willmcgugan — APPROVED** [review]
